Add options to limit write buffer and modify checksum behavior

WallaceIT opened the pull request:
Add command line options to:

- limit write buffer size, useful on systems with limited RAM memory but large flash device(s)
- modify the checksum behavior, either verifying the written data instead of the source one or skipping the checksum entirely

WallaceIT committed:
Store ranges in numerical form

Signed-off-by: Francesco Valla <valla.francesco@gmail.com>

WallaceIT committed:
Split checksum calculation in multiple functions

A single checksum function does not allow to calculate the checksum in
multiple steps, as it is required if a write is divided in steps.

Signed-off-by: Francesco Valla <valla.francesco@gmail.com>

embetrix commented on README.md:
````diff
@@ -22,12 +22,20 @@ Unlike the Yocto BMAP tool, `bmap-writer` is C++ based does not require Python a
+bmap-writer [-b <max-buf-size>] [-v] <image-file> <bmap-file> <target-device>
 ```
 
+The `<max-buf-size>` parameter can be used to limit the write buffer size on systems with limited RAM.
````
would be possible to calculate the available free memory instead ?

for example using sysinfo : https://man7.org/linux/man-pages/man2/sysinfo.2.html

I would like to keep the tool as simple as possible and not introduce much options (or not at all)

WallaceIT replied:
Yes, that can be an option, but if during execution the free memory lowers significantly, memory exhaustion can still be a risk.

Since the tool is meant for the embedded world, having a non-mandatory option that can be used to fix an arbitrary maximum (thus allowing for example parallel writes on different flash devices) is IMO still useful.

The two approaches can also be used together: if the user does not specify manually a maximum value, the tool can auto-calculate it itself from free memory.

WallaceIT replied:
I inserted the query of sysinfo in the updated MR - let me know what you think. Thjank you

embetrix commented on README.md:
```diff
@@ -22,12 +22,20 @@ Unlike the Yocto BMAP tool, `bmap-writer` is C++ based does not require Python a
+It can be specified as a plain number or using the `K` (KiB), `M` (MiB), `G` (GiB) or `T` (TiB) suffixes.
+If not specified, the write buffer size will be variable and will depend on the size of the ranges specified inside the BAMP file.
+
+If `-n` is specified, the checksum verification is skipped.
```
same here can we keep the checksum verification enforced ?

WallaceIT replied:
I introduced the `n` option to mimic the feature provided by the original bmatools, but it may be removed. It's a non-mandatory option though, if not specified the default behavior will be kept.

WallaceIT replied:
I removed the skip option entirely in the updated MR.

embetrix commented on README.md:
```diff
@@ -22,12 +22,20 @@ Unlike the Yocto BMAP tool, `bmap-writer` is C++ based does not require Python a
+
+If `-n` is specified, the checksum verification is skipped.
+If `-w` is specified, the written data is read back and verified against the checksum contained inside the BMAP file.
+Otherwise, the checksum verification is performed on the data read from the image file.
```
this is a good thing to check the written checksum instead of the input one

WallaceIT replied:
Yes, even if OS caching could still be in the middle

embetrix replied:
can we have only checksum of written date or skipping it option ? the read checksum option is redundnant

embetrix commented on bmap-writer.cpp:
```diff
@@ -26,29 +26,69 @@
+    CHECKSUM_WRITTEN,
+};
+
+class Checksummer {
```
I would like to avoid using class object oriented code here.

WallaceIT replied:
Ok, I'll try to rewrite it using a `struct` + init/deinit functions.

WallaceIT replied:
Done in updated MR.

embetrix commented on bmap-writer.cpp:
```diff
@@ -123,34 +146,18 @@ bool isDeviceMounted(const std::string &device) {
 
     try {
         size_t decHead = 0;
+        int rwMode = (checksumMode == CHECKSUM_WRITTEN) ? O_RDWR : O_WRONLY;
```
I think we can open it always in RW mode there should be no major drawbacks on performance

WallaceIT replied:
Ok

embetrix commented on bmap-writer.cpp:
```diff
@@ -123,34 +146,18 @@ bool isDeviceMounted(const std::string &device) {
+        int rwMode = (checksumMode == CHECKSUM_WRITTEN) ? O_RDWR : O_WRONLY;
 
-        dev_fd = open(device.c_str(), O_WRONLY | O_CREAT | O_SYNC, S_IRUSR | S_IWUSR);
+        dev_fd = open(device.c_str(), rwMode | O_CREAT | O_SYNC, S_IRUSR | S_IWUSR);
```
a consitent naming convention should be used : O_RDWR

WallaceIT replied:
Yes, if `O_RDWR` is the only flag used.

embetrix commented on bmap-writer.cpp:
```diff
@@ -26,29 +26,83 @@
     size_t blockSize;
 };
 
+enum ChecksumMode {
```
I find the checksum refactoring too much ? what kind of benefits does does bring ?

WallaceIT replied:
Configurability!

Nevertheless, I got your point on the tool being minimalistic and remove this completely.

embetrix commented on bmap-writer.cpp:
```diff
@@ -185,66 +206,106 @@ int BmapWriteImage(const std::string &imageFile, const bmap_t &bmap, const std::
+
+            //std::cout << "Processing Range: startBlock=" << range.startBlock << ", endBlock=" << range.endBlock << std::endl;
+
+            while ((writtenSize < rangeSize) && !endOfFile) {
```
can't we have at each range loop a calculation of available free memory and adjust it accordingly ? thus we can avoid at all the parameters _maxbuffersize_ ?

WallaceIT replied:
Done in updated MR.

WallaceIT committed:
Compute maximum buffer size before writing each range

To avoid exhausting the available RAM, query the free amount of it
before allocating the write buffer and use the half of this value as
maximum buffer size.

If required (i.e., if the range size is greater than the maximum
buffer size), split write and checksum calculation in multiple passes.

Signed-off-by: Francesco Valla <valla.francesco@gmail.com>

embetrix commented:
hello @WallaceIT , I made some small modification to simplify SHA256 calculation and set the checksum verification for the written buffer instead, can you please rebase/adapt your MR: 

* adding option `-n` to ignore/skip sha256 calculation
* dynamically calculate write buffer size at each iteration (without having an option for <max-buf-size>)

Thank you !

WallaceIT commented:
Hello @embetrix, I'll rebase the MR, but be advised that I'll need to return to the previous implementation of the SHA256 function(s), since the hash will potentially be calculated in multiple steps.

embetrix commented:
@WallaceIT thanks, is it possible to use SHA256_Init/SHA256_Update/SHA256_Final instead ?

what do you think about maxBufferSize ? should be  we use 1/2 of available memory ?

WallaceIT committed:
Add -n option to skip checksum verification

Signed-off-by: Francesco Valla <valla.francesco@gmail.com>

WallaceIT commented:
> @WallaceIT thanks, is it possible to use SHA256_Init/SHA256_Update/SHA256_Final instead ?

Are you sure? They are currently deprecated, see https://docs.openssl.org/3.2/man3/SHA256_Init/

> what do you think about maxBufferSize ? should be we use 1/2 of available memory ?

Maybe yes, even if I still believe that offering the user the possibility to set its own size would be useful :)
What do you think of a dual solution? I.e.: half of the available memory (calculated at each cycle) if nothing is specified, otherwise user-specified amount?

embetrix commented:
Thanks for the work :+1:  yes those SHA functions are deprecated, I was thinking about not using OpenSSL but a header only SHA256 lib : https://github.com/imahjoub/hash_sha256.

User Specified amount could be a bit misleading for inexperienced User. I think the way you implemented now should be transparen

WallaceIT commented:
> Thanks for the work 👍 yes those SHA functions are deprecated, I was thinking about not using OpenSSL but a header only SHA256 lib : https://github.com/imahjoub/hash_sha256.

+1 for this approach, as it would reduce the size of the dependencies (right now OpenSSL is quite big)


Thank you!
